oauth2device/oauth2device_test.go

```diff
@@ -0,0 +1,274 @@
+	_, err := c.Wait(context.Background(), new(DeviceCode))
+	urlErr := &url.Error{Op: "parse", URL: "this://is\nwrong", Err: errors.New("net/url: invalid control character in URL")}
+	if got, want := err, urlErr; !reflect.DeepEqual(got, want) {
+		t.Errorf("incorrect error;\n\t got %q\n\twant %q", got.Error(), want.Error())
+	}
+}
+
+func TestWait(t *testing.T) {
+	tests := []struct {
+		name  string
+		resps []string
+		tok   *oauth2.Token
+		err   error
+	}{
+		{
+			name:  "invalid json response",
+			resps: []string{`this{\nisn't good`},
+			tok:   nil,
+			err:   ErrFetchToken,
+		},
+		{
+			name:  "valid token",
+			resps: []string{`{"access_token":"foo","id_token":"bar"}`},
+			tok: (&oauth2.Token{AccessToken: "foo"}).WithExtra(
+				map[string]interface{}{"id_token": "bar"},
+			),
+			err: nil,
+		},
+		{
+			name: "authorization pending",
+			resps: []string{
+				`{"error":"authorization_pending"}`,
+				`{"access_token":"foo","id_token":"bar"}`,
+			},
+			tok: (&oauth2.Token{AccessToken: "foo"}).WithExtra(
+				map[string]interface{}{"id_token": "bar"},
+			),
+			err: nil,
+		},
+		{
+			name: "slow down",
+			resps: []string{
+				`{"error":"slow_down"}`,
+				`{"access_token":"foo","id_token":"bar"}`,
+			},
+			tok: (&oauth2.Token{AccessToken: "foo"}).WithExtra(
+				map[string]interface{}{"id_token": "bar"},
+			),
+			err: nil,
+		},
+		{
+			name:  "access denied",
+			resps: []string{`{"error":"access_denied"}`},
+			tok:   nil,
+			err:   ErrAccessDenied,
+		},
+		{
+			name:  "other error",
+			resps: []string{`{"error":"what in the!?"}`},
+			tok:   nil,
+			err:   errors.New("authorization failed: what in the!?"),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			var count int
+			srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				fmt.Fprint(w, test.resps[count])
+				count++
+			}))
+			defer srv.Close()
+
+			c := Config{
+				Config: &oauth2.Config{
+					Endpoint: oauth2.Endpoint{TokenURL: srv.URL},
+				},
+			}
+
+			tok, err := c.Wait(context.Background(), new(DeviceCode))
+			if got, want := err, test.err; !reflect.DeepEqual(got, want) {
+				t.Errorf("incorrect error;\n\t got %q\n\twant %q", got.Error(), want.Error())
+				t.Logf("err: %#v", got)
+			}
+
+			if got, want := tok, test.tok; !reflect.DeepEqual(got, want) {
+				t.Errorf("incorrect token;\n\t got %#v\n\twant %#v", got, want)
+			}
+		})
+	}
+}
+
+type errorRoundTripper struct{}
+
+func (e errorRoundTripper) RoundTrip(*http.Request) (*http.Response, error) {
+	return nil, errors.New("nope")
+}
+
+var errorClient = &http.Client{Transport: errorRoundTripper{}}
+
+func staticHandler(s string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprint(w, s)
+	})
+}
+
+/*
+func (c *Config) DeviceCode(ctx context.Context) (*DeviceCode, error) {
+func (c *Config) Wait(ctx context.Context, code *DeviceCode) (*oauth2.Token, error) {
+*/
```

Comment:
Are these needed for something?

Reply:
Ah, good catch. I put sigs in the test file as a reference as I'm writing tests and I forgot to remove these.

Reply:
makes sense :grin: